File: vjudge/2020七月暑假集训/Nowcoder 多校 2020/Day 5/E.cc
/**
 *
 * WA 了，是需要 __int128 吗
 */
#include <iostream>
#include <cstdio>
#include <cctype>
#include <bitset>
#include <vector>
#include <algorithm>
#include <cstring>
#include <cmath>

using namespace std;
using longs = long long;
using uint = unsigned;
using decimal = __int128;

inline int nextInt()
{
    int x = 0, f = 1, ch = getchar();
    while (!isdigit(ch)) if (ch == '-') f = -1, ch = getchar();
    while (isdigit(ch)) x = x * 10 + ch - 48, ch = getchar();
    return x * f;
}

ostream &operator <<(ostream &out, __int128 x)
{
    if (x < 0) { out << '-'; x = -x; }
    static char str[40]; int cnt = 0;
    while (x > 9) { str[cnt ++] = (x % 10) ^ 48; x /= 10;}
    str[cnt ++] = x ^ 48;
    while (cnt --) out << str[cnt];
    return out;
}

inline decimal gcd(decimal a, decimal b)
{
    if (a < b) swap(a, b);
    if (!b) return a;
    else return gcd(b, a % b);
}

inline decimal lcm(decimal a, decimal b)
{ return a * b / gcd(a, b); }

const int N = 1e5 + 5;
int p[N], n;
bitset<N> vis;
vector<int> wa;

int main()
{
    ios::sync_with_stdio(false);
    std::cin.tie(nullptr);
    std::cout.tie(nullptr);

    cin >> n;
    wa.clear();
    for (int i = 1; i <= n; ++ i) cin >> p[i];
    for (int i = 1; i <= n; ++ i)
        if (!vis[i])
        {
            vis[i] = true;
            int cnt = 1, cc = i;
            while (!vis[p[cc]])
                cc = p[cc], vis[cc] = true, ++ cnt;
            wa.push_back(cnt);
        }
    decimal ans = 1;
    for (auto & ii : wa) ans = lcm(ans, ii);
    auto mod = (decimal)pow(10, n);
    if (n <= 36) ans %= mod;
    cout << ans << endl;

    return 0;
}

File: vjudge/2020七月暑假集训/Nowcoder 多校 2020/Day 5/E.java
/*
 * 确实，并不需要考虑什么取模，根本就不会炸
 * 这波啊，这波考的是睾精度！
 *
 * 你也该熟悉熟悉 Java 算法题的写法了
 * 没想到 Java 竟然自带 gcd 方法，这有点方便
 * 
 * 牛客的 Javac 竟然不支持 var，啊这 ==
 */
import java.math.BigInteger;
import java.util.*;

public class Main
{
    final static int N = (int)(1e5 + 5);
    final static BigInteger zero = new BigInteger("0");

    public static BigInteger gcd(BigInteger a, BigInteger b)
    {
        if (a.compareTo(b) < 0) return gcd(b, a);
        if (b.equals(zero)) return a;
        else return gcd(b, a.mod(b));
    }

    public static void main(String[] args)
    {
        var p = new int[N];
        var vis = new boolean[N];
        var sc = new Scanner(System.in);
        var ans = BigInteger.valueOf(1);
        List<BigInteger> cir = new Vector<>();
        int n = sc.nextInt();

        for (int i = 1; i <= n; ++ i)
            p[i] = sc.nextInt();
        for (int i = 1; i <= n; ++ i)
            if (!vis[i])
            {
                vis[i] = true;
                int cnt = 1, cc = i;
                while (!vis[p[cc]])
                {
                    cc = p[cc];
                    vis[cc] = true;
                    ++ cnt;
                }
                if (cnt > 1) cir.add(BigInteger.valueOf(cnt));
            }
        for (var ii : cir)
            ans = ans.multiply(ii.divide(ii.gcd(ans)));
        System.out.println(ans.toString());
    }
}
